**SA367 &#x25aa; Mathematical Models for Decision Making &#x25aa; Spring 2022 &#x25aa; Uhan**

# Lesson 4. The Mileage Running Problem

## The problem

Professor May B. Wright needs to fly from Baltimore (BWI) to Los Angeles (LAX) to attend a conference.
She thinks this would be the perfect opportunity to accumulate some frequent flyer miles on American Airlines (AA), where she already has Platinum status.

Looking into flights on AA, she sees that every itinerary from BWI to LAX costs roughly the same.
She has a full day to spare for travel, so she wants to know: which sequence of AA domestic flights starting at BWI and ending at LAX over the course of one day will allow her to accumulate the most miles?

* Yes, people actually do this: this is known as __mileage running__
    - Apparently, this has become harder to do over the past decade
    - [An older article from the New York Times](https://www.nytimes.com/2014/09/14/upshot/the-fadeout-of-the-mileage-run.html)
    - [An even older article from Wired](https://www.wired.com/2007/07/mileage-runner/)

## Modeling the problem

* Suppose we have a database of every AA domestic flight on a given day


* In particular, for each flight, we have:
    - the flight number
    - the origin airport
    - the destination airport
    - the departure time at the origin airport
    - the arrival time at the destination airport
    - the distance traveled in miles


* How can we formulate Professor Wright's problem as a shortest path problem?

## pandas (the package, not the animal 🐼)

* In the same folder as this notebook, there is a file called `aa_domestic_flights.csv` with the database described above


* `.csv` stands for __comma-separated values__


* We can view `.csv` files in Excel


* We can also view `.csv` files directly in JupyterLab &mdash; let's take a look!

* We can use this data with Pythjon using [__pandas__](https://pandas.pydata.org/), a Python package for data manipulation and analysis
    - It's especially useful for cleaning and manipulating datasets


* pandas does a lot of stuff &mdash; [here is the official documentation for pandas](http://pandas.pydata.org/pandas-docs/stable/index.html)
    

* In this lesson, we'll use pandas in a very basic way to set up the shortest path problem we formulated above

* To use pandas, we first need to import it, like this:

In [1]:
import pandas as pd

* A pandas __DataFrame__ is just a two-dimensional table, with rows and columns

* We can use the `pd.read_csv()` function in pandas to read `aa_domestic_flights.csv` into a DataFrame called `df`, like this:

In [2]:
# Read csv file into a DataFrame
# Designate departure and arrival time columns as dates
df = pd.read_csv('aa_domestic_flights.csv', parse_dates=['DEP_TIME', 'ARR_TIME'])

* By default, `pd.read_csv()` assumes the first row of the csv file contains the names of each column


* The `parse_dates` argument tells `pd.read_csv()` which columns correspond to dates, so that we can perform date-specific calculations on these columns later


* [Here is the official documentation for `pd.read_csv()`](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html)

* It's a good idea to take a quick look at the DataFrame `pd.read_csv()` creates, just in case something went wrong


* To examine the first 5 rows of a DataFrame, we can use the `.head()` method:

In [3]:
# Solution
df.head()

FLIGHT ORIGIN DEST            DEP_TIME            ARR_TIME  DISTANCE
0    1-BOS-JFK    BOS  JFK 2016-09-01 06:00:00 2016-09-01 07:15:00     187.0
1   40-BOS-ORD    BOS  ORD 2016-09-01 19:12:00 2016-09-01 22:02:00     867.0
2  147-BOS-LAX    BOS  LAX 2016-09-01 15:15:00 2016-09-01 21:45:00    2611.0
3  197-BOS-ORD    BOS  ORD 2016-09-01 15:30:00 2016-09-01 18:24:00     867.0
4  198-BOS-JFK    BOS  JFK 2016-09-01 13:10:00 2016-09-01 14:31:00     187.0

- The first column is the __index__ of the DataFrame
    - The index provides a label for each row of the DataFrame


- Right now, the index is sort of uninformative


- Since each row corresponds to a flight, it would be nice if the index corresponded to the flight names


- We can do this using the `.set_index()` method:

In [4]:
# Solution
df = df.set_index("FLIGHT")

- Let's see what this accomplished by printing the first 5 rows of the DataFrame `df` again:

In [5]:
# Solution
df.head()

ORIGIN DEST            DEP_TIME            ARR_TIME  DISTANCE
FLIGHT                                                                   
1-BOS-JFK      BOS  JFK 2016-09-01 06:00:00 2016-09-01 07:15:00     187.0
40-BOS-ORD     BOS  ORD 2016-09-01 19:12:00 2016-09-01 22:02:00     867.0
147-BOS-LAX    BOS  LAX 2016-09-01 15:15:00 2016-09-01 21:45:00    2611.0
197-BOS-ORD    BOS  ORD 2016-09-01 15:30:00 2016-09-01 18:24:00     867.0
198-BOS-JFK    BOS  JFK 2016-09-01 13:10:00 2016-09-01 14:31:00     187.0

* A column by itself is called a **Series**


* You can select the Series `DEST` of the DataFrame `df` like this:

    ```python
    df["DEST"]
    ```


* So, to print the Series `DEST`, we could write:

In [6]:
# Solution
print(df["DEST"])

FLIGHT
1-BOS-JFK       JFK
40-BOS-ORD      ORD
147-BOS-LAX     LAX
197-BOS-ORD     ORD
198-BOS-JFK     JFK
               ... 
621-BOI-DFW     DFW
2357-LBB-DFW    DFW
2571-ANC-DFW    DFW
483-GEG-PHX     PHX
490-GEG-PHX     PHX
Name: DEST, Length: 2607, dtype: object


## Making the data easier to use

- We have successfully imported our data into Python! 


- We could set up our shortest path problem using the DataFrame directly, but this can be a bit cumbersome


- Let's take some additional steps that will make setting up our shortest path problem a bit easier

- First, let's get a list of the flights


- This will be useful, since the nodes in our shortest path problem correspond to the flights


- In the DataFrame `df` we defined above, the index consists of the flights


- We can get a list of the index values of `df` with `list(df.index.values)`, like this:

In [7]:
# Solution
flights = list(df.index.values)

- Let's check our work and inspect `flights` by printing it:

In [8]:
# Solution
print(f"Flights: {flights}")

Flights: ['1-BOS-JFK', '40-BOS-ORD', '147-BOS-LAX', '197-BOS-ORD', '198-BOS-JFK', '85-BOS-JFK', '252-BOS-ORD', '333-BOS-LAX', '1086-BOS-MIA', '1094-BOS-DFW', '307-BOS-PHX', '1155-BOS-ORD', '1172-BOS-MIA', '1211-BOS-DFW', '1274-BOS-MIA', '166-BOS-LAX', '175-BOS-DFW', '1435-BOS-ORD', '1404-BOS-ORD', '1503-BOS-ORD', '1509-BOS-MIA', '1240-BOS-ORD', '223-BOS-LAX', '1006-BOS-MIA', '1039-BOS-JFK', '2303-BOS-DFW', '2251-BOS-DFW', '2253-BOS-ORD', '2378-BOS-PHX', '2454-BOS-MIA', '2552-BOS-DFW', '421-BOS-PHL', '1707-BOS-PHL', '1787-BOS-CLT', '1788-BOS-CLT', '1805-BOS-CLT', '1879-BOS-CLT', '1907-BOS-PHX', '1925-BOS-CLT', '2068-BOS-CLT', '2116-BOS-DCA', '2118-BOS-LGA', '2119-BOS-DCA', '2120-BOS-DCA', '2121-BOS-LGA', '2124-BOS-DCA', '2126-BOS-LGA', '2127-BOS-DCA', '2129-BOS-DCA', '2130-BOS-DCA', '2131-BOS-LGA', '2134-BOS-DCA', '2136-BOS-LGA', '2139-BOS-DCA', '2141-BOS-LGA', '2142-BOS-DCA', '2146-BOS-LGA', '2148-BOS-LGA', '2149-BOS-DCA', '2152-BOS-LGA', '2158-BOS-LGA', '2159-BOS-DCA', '2160-BOS-DCA',

* While we're here, let's make sure we have the right number of flights in the variable `flights`:

In [9]:
# Solution
print(f"Number of flights: {len(flights)}")

Number of flights: 2607


- We also want to easily access the origin, destination, departure time, arrival time, and distance for each flight


- These correspond to the columns in our DataFrame `df`


- We can convert DataFrame columns to dictionaries as using the `.to_dict()` on the column of interest, like this:

In [10]:
# Solution
origin = df['ORIGIN'].to_dict()
destination = df['DEST'].to_dict()
departure_time = df['DEP_TIME'].to_dict()
arrival_time = df['ARR_TIME'].to_dict()
distance = df['DISTANCE'].to_dict()

- As a result, we can access information about each flight through these dictionaries as follows:

In [ ]:
# Information about flight 1240-BOS-ORD
print(f"Origin: { }")
print(f"Destination: { }")
print(f"Departure time: { }")
print(f"Arrival time: { }")
print(f"Distance: { }")

In [11]:
# Solution
# Information about flight 1240-BOS-ORD
print(f"Origin: {origin['1240-BOS-ORD']}")
print(f"Destination: {destination['1240-BOS-ORD']}")
print(f"Departure time: {departure_time['1240-BOS-ORD']}")
print(f"Arrival time: {arrival_time['1240-BOS-ORD']}")
print(f"Distance: {distance['1240-BOS-ORD']}")

Origin: BOS
Destination: ORD
Departure time: 2016-09-01 12:15:00
Arrival time: 2016-09-01 15:05:00
Distance: 867.0


## Setting up the shortest path problem in NetworkX 

* Now we're ready to set up the shortest path problem we formulated above

* First, let's import `networkx` and `bellmanford` so we can use them:

In [12]:
import networkx as nx
import bellmanford as bf

### Adding nodes

* Let's build the shortest path graph, starting with an empty directed graph:

In [13]:
# Solution
G = nx.DiGraph()

* Next, let's create a "start" and "end" node

In [14]:
# Solution
G.add_node("start")
G.add_node("end")

* Now, we need to add a node for each flight, or each row of our database

In [15]:
# Solution
for flight in flights:
    G.add_node(flight)

* The `.number_of_nodes()` method applied to a NetworkX graph &mdash; well, you can guess what it does


* Or, you can just try it out:

In [16]:
# Print number of nodes in G
print(G.number_of_nodes())

2609


### Adding edges

* Now we can go over every pair of flight nodes, and check if we need to add an edge between them
    - Remember the length of these edges is the <span class="rred">negative</span> of the distance of the first flight


* To add or subtract times, we need to use `pd.to_timedelta()` &mdash; [here is the documentation](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_timedelta.html)
    - For example, to subtract 30 minutes, we would write 
    ```python
    some_time_variable - pd.to_timedelta(30, unit="m")
    ```


* This might seem awkward, but if you think about it, working with dates and time _is_ awkward &mdash; you need to keep track of different (non-base-10) units

In [ ]:
# Iterate through every pair of flight nodes
        
        # If the first flight arrives where the second flight departs...
            
            # And if the first flight arrives 45 minutes before the second flight leaves,
            # add an edge from the first flight to the second


In [17]:
# Solution
# Iterate through every pair of flight nodes
for first in flights:
    for second in flights:
        
        # If the first flight arrives where the second flight departs...
        if (destination[first] == origin[second]):
            
            # And if the first flight arrives 45 minutes before the second flight leaves,
            # add an edge from the first flight to the second
            if (arrival_time[first] + pd.to_timedelta(45, unit="m") < departure_time[second]):
                G.add_edge(first, second, length=-distance[first])

* Finally, we need to add edges:
    - from the start node to all flights departing from BWI
    - from all flights arriving at LAX to the end node

In [ ]:
# Iterate through all flights

    # If the flight departs from BWI, 
    # add an edge from start to this flight
        
    # If the flight arrives at LAX, 
    # add an edge from this flight to end


In [18]:
# Solution
# Iterate through all flights
for flight in flights:

    # If the flight departs from BWI, 
    # add an edge from start to this flight
    if origin[flight] == "BWI":
        G.add_edge("start", flight, length=0)
        
    # If the flight arrives at LAX, 
    # add an edge from this flight to end
    if destination[flight] == "LAX":
        G.add_edge(flight, "end", length=-distance[flight])

* Similar to `G.number_of_nodes()`, we can perform a sanity check with our work with `G.number_of_edges()`

In [19]:
print(G.number_of_edges())

158335


## Solving the shortest path problem, interpreting the output

* Now that we have our directed graph set up, we can solve for the shortest path from the start node to the end node just like we did in the last lesson:

In [20]:
# Solve the shortest path problem using Bellman-Ford
length, nodes, negative_cycle = bf.bellman_ford(G, source="start", target="end", weight="length")

# Print output from Bellman-Ford
print(f"Negative cycle? {negative_cycle}")
print(f"Shortest path length: {length}")
print(f"Shortest path: {nodes}")

Negative cycle? False
Shortest path length: -8005.0
Shortest path: ['start', '1817-BWI-CLT', '658-CLT-LAS', '1584-LAS-PHX', '694-PHX-HNL', '298-HNL-LAX', 'end']


* What does the output tell us about how to solve Professor Wright's problem?

_Write your notes here. Double-click to edit._

_Solution._

- The length of the shortest path is the negative of the maximum total distance Professor Wright can travel from BWI to LAX. In this case, the maximum total distance is 8005 miles.

- The nodes in the shortest path tells us which flights Professor Wright should take:
    - 1817 from BWI to CLT
    - 658 from CLT to LAS
    - 1584 from LAS to PHX
    - 694 from PHX to HNL
    - 298 from HNL to LAX

## On your own...

Suppose Professor Wright wants to find the longest itinerary from IAD (Washington DC - Dulles) to SAN (San Diego) instead.

In the cell below, write the code that sets up and solves the shortest path formulation for her problem from start to finish.

In the cell after, describe in words what the output from the Bellman-Ford algorithm tells you about how to solve Professor Wright's problem.

In [21]:
# Solution
# Import packages
import pandas as pd
import networkx as nx
import bellmanford as bf

# Read csv file into a DataFrame
# Designate departure and arrival time columns as dates
df = pd.read_csv('aa_domestic_flights.csv', parse_dates=['DEP_TIME', 'ARR_TIME'])

# Create empty networkx digraph
G = nx.DiGraph()

# Create start and end nodes
G.add_node("start")
G.add_node("end")

# Add a node for each flight
for flight in flights:
    G.add_node(flight)
    
# Iterate through every pair of flight nodes
for first in flights:
    for second in flights:
        
        # If the first flight arrives where the second flight departs...
        if (destination[first] == origin[second]):
            
            # And if the first flight arrives 45 minutes before the second flight leaves,
            # add an edge from the first flight to the second
            if (arrival_time[first] + pd.to_timedelta(45, unit="m") < departure_time[second]):
                G.add_edge(first, second, length=-distance[first])

# Iterate through all flights
for flight in flights:

    # If the flight departs from IAD, 
    # add an edge from start to this flight
    if origin[flight] == "IAD":
        G.add_edge("start", flight, length=0)
        
    # If the flight arrives at SAN, 
    # add an edge from this flight to end
    if destination[flight] == "SAN":
        G.add_edge(flight, "end", length=-distance[flight])
        
# Solve the shortest path problem using Bellman-Ford
length, nodes, negative_cycle = bf.bellman_ford(G, source="start", target="end", weight="length")

# Print output from Bellman-Ford
print(f"Negative cycle? {negative_cycle}")
print(f"Shortest path length: {length}")
print(f"Shortest path: {nodes}")

Negative cycle? False
Shortest path length: -6005.0
Shortest path: ['start', '2636-IAD-LAX', '2503-LAX-ORD', '2375-ORD-DFW', '435-DFW-SAN', 'end']


_Write your notes here. Double-click to edit._

_Solution._

- The length of the shortest path is the negative of the maximum total distance Professor Wright can travel from IAD to SAN. In this case, the maximum total distance is 6005 miles.

- The nodes in the shortest path tells us which flights Professor Wright should take:
    - 2636 from IAD to LAX
    - 2503 from LAX to ORD
    - 2375 from ORD to DFW
    - 435 from DFW to SAN